File: CSharp/Client/ToProtobuf.cs
#region

using System;
using System.Collections.Generic;
using System.IO;
using ProtoBuf;
using ThingModel.Proto;

#endregion

namespace ThingModel.Client
{
    public class ToProtobuf
    {
        // String declarations dictionary, in the sender side
        protected readonly IDictionary<string, int> StringDeclarations = new Dictionary<string, int>();

        // Current transaction object
        // It will be filled step by step
        protected Transaction Transaction;

        // Simple binding between ThingModel property types and protocol buffer property types
        // Tested as alternative to the dynamic keyword
        private readonly Dictionary<Type, Proto.PropertyType.Type> _prototypesBinding = new Dictionary
            <Type, Proto.PropertyType.Type>
            {
                {typeof (Property.Location), Proto.PropertyType.Type.LOCATION},
                {typeof (Property.String), Proto.PropertyType.Type.STRING},
                {typeof (Property.Double), Proto.PropertyType.Type.DOUBLE},
                {typeof (Property.Int), Proto.PropertyType.Type.INT},
                {typeof (Property.Boolean), Proto.PropertyType.Type.BOOLEAN},
                {typeof (Property.DateTime), Proto.PropertyType.Type.DATETIME},
            };

        // When a string is in this collection, it should be sended
        // as an int key with the StringToKey method
        private readonly HashSet<string> _stringToDeclare = new HashSet<string>();

        private readonly Dictionary<int, Proto.Thing> _thingsState = new Dictionary<int, Proto.Thing>();
        private readonly Dictionary<Tuple<int, int>, Proto.Property> _propertiesState = new Dictionary<Tuple<int, int>, Proto.Property>();

        protected int StringToKey(string value)
        {
            if (String.IsNullOrEmpty(value))
            {
                return 0;
            }

            int key;
            if (StringDeclarations.TryGetValue(value, out key))
            {
                return key;
            }

            key = StringDeclarations.Count + 1;

            StringDeclarations[value] = key;

            Transaction.string_declarations.Add(new StringDeclaration
                {
                    key = key,
                    value = value
                });

            return key;
        }

        public Transaction Convert(IEnumerable<Thing> publish, IEnumerable<Thing> delete,
                                   IEnumerable<ThingType> declarations, string senderID)
        {
            Transaction = new Transaction();

            Transaction.string_sender_id = StringToKey(senderID);

            foreach (var thing in publish)
            {
                Convert(thing);
            }

            ConvertDeleteList(delete);
            ConvertDeclarationList(declarations);

            return Transaction;
        }


        private readonly MemoryStream _memoryInput = new MemoryStream();

        public byte[] Convert(Transaction transaction)
        {
            _memoryInput.SetLength(0);
            Serializer.Serialize(_memoryInput, transaction);
            return _memoryInput.ToArray();
        }

        protected Proto.Thing Convert(Thing thing)
        {
            var change = false;

            var publication = new Proto.Thing
                {
                    string_id = StringToKey(thing.ID),
                    string_type_name = thing.Type != null ? StringToKey(thing.Type.Name) : 0,
                    connections_change = false
                };
            
            Proto.Thing previousThing;
            _thingsState.TryGetValue(publication.string_id, out previousThing);

            if (previousThing == null)
            {
                change = true;
            }

            foreach (var connectedThing in thing.ConnectedThings)
            {
                var connectionKey = StringToKey(connectedThing.ID);
                publication.connections.Add(connectionKey);

                if (previousThing == null || !previousThing.connections.Contains(connectionKey))
                {
                    publication.connections_change = true;
                }
            }

            if (previousThing != null && publication.connections_change == false)
            {
                if (previousThing.connections.Count != publication.connections.Count)
                {
                    publication.connections_change = true;
                }
                else
                {
                    foreach (var key in previousThing.connections)
                    {
                        if (!publication.connections.Contains(key))
                        {
                            publication.connections_change = true;
                            break;
                        }
                    }
                }
            }
            
            // If we don't have changes on the connection list,
            // it's useless to send it
            if (!publication.connections_change)
            {
                if (publication.connections.Count > 0)
                {
                    publication.connections.Clear();    
                }
                
            }
            else
            {
                change = true;
            }

            foreach (var property in thing.GetProperties())
            {
                var proto = new Proto.Property
                    {
                        string_key = StringToKey(property.Key)
                    };

                ConvertProperty((dynamic) property, proto);

                var key = new Tuple<int, int>(publication.string_id, proto.string_key);

                if (previousThing != null)
                {
                    Proto.Property previousProto;
                    if (_propertiesState.TryGetValue(key, out previousProto))
                    {
                        // ReSharper disable CompareOfFloatsByEqualityOperator
                        // Boring but necessary ?
                        if (previousProto.type == proto.type &&
                            previousProto.boolean_value == proto.boolean_value &&
                            previousProto.datetime_value == proto.datetime_value &&
                            previousProto.double_value == proto.double_value &&
                            previousProto.int_value == proto.int_value &&
                            ((previousProto.location_value == null && proto.location_value == null) ||
                             (previousProto.location_value != null && proto.location_value != null &&
                              previousProto.location_value.x == proto.location_value.x &&
                              previousProto.location_value.y == proto.location_value.y &&
                              previousProto.location_value.z == proto.location_value.z &&
                              previousProto.location_value.z_null == proto.location_value.z_null &&
                              previousProto.location_value.string_system == proto.location_value.string_system)) &&
                            ((previousProto.string_value == null && proto.string_value == null) ||
                             (previousProto.string_value != null && proto.string_value != null &&
                              previousProto.string_value.string_value == proto.string_value.string_value &&
                              previousProto.string_value.value == proto.string_value.value)))
                        {
                            continue;
                        }

                        // ReSharper restore CompareOfFloatsByEqualityOperator
                                    
                    }

                    
                }

                change = true;
                
                publication.properties.Add(proto);
                _propertiesState[key] = proto;
            }

            if (change)
            {
                Transaction.things_publish_list.Add(publication);
                if (previousThing == null)
                {
                    _thingsState.Add(publication.string_id, publication);
                }
            }
            

            return publication;
        }

        protected void ConvertDeleteList(IEnumerable<Thing> publish)
        {
            foreach (var thing in publish)
            {
                var key = StringToKey(thing.ID);
                Transaction.things_remove_list.Add(key);
                _thingsState.Remove(key);
            }
        }

        protected void ConvertDeclarationList(IEnumerable<ThingType> declarations)
        {
            foreach (var thingType in declarations)
            {
                var declaration = new Proto.ThingType
                    {
                        string_name = StringToKey(thingType.Name),
                        description = thingType.Description
                    };

                foreach (var propertyType in thingType.GetProperties())
                {
                    declaration.properties.Add(new Proto.PropertyType
                        {
                            string_key = StringToKey(propertyType.Key),
                            name = propertyType.Name,
                            description = propertyType.Description,
                            required = propertyType.Required,
                            type = _prototypesBinding[propertyType.Type]
                        });
                }

                Transaction.thingtypes_declaration_list.Add(declaration);
            }
        }

        protected void ConvertProperty(Property.Location property, Proto.Property proto)
        {
            var value = property.Value;

            if (value is Location.LatLng)
            {
                proto.type = Proto.Property.Type.LOCATION_LATLNG;
            }
            else if (value is Location.Point)
            {
                proto.type = Proto.Property.Type.LOCATION_POINT;
            }
            else
            {
                proto.type = Proto.Property.Type.LOCATION_EQUATORIAL;    
            }
            
            proto.location_value = new Proto.Property.Location
                {
                    x = value.X,
                    y = value.Y,
                    string_system = StringToKey(value.System),
                    z_null = value.Z == null
                };

            if (value.Z != null)
            {
                proto.location_value.z = (double) value.Z;
            }
        }

        protected void ConvertProperty(Property.String property, Proto.Property proto)
        {
            var value = property.Value;

            proto.type = Proto.Property.Type.STRING;
            proto.string_value = new Proto.Property.String();
            
            if (_stringToDeclare.Contains(value))
            {
                proto.string_value.string_value = StringToKey(value);
            }
            else
            {
                proto.string_value.value = value;

                // Use string to key the next time
                _stringToDeclare.Add(value);
            }
        }

        protected void ConvertProperty(Property.Double property, Proto.Property proto)
        {
            proto.type = Proto.Property.Type.DOUBLE;
            proto.double_value = property.Value;
        }

        protected void ConvertProperty(Property.Int property, Proto.Property proto)
        {
            proto.type = Proto.Property.Type.INT;
            proto.int_value = property.Value;
        }

        protected void ConvertProperty(Property.Boolean property, Proto.Property proto)
        {
            proto.type = Proto.Property.Type.BOOLEAN;
            proto.boolean_value = property.Value;
        }

        protected void ConvertProperty(Property.DateTime property, Proto.Property proto)
        {
            proto.type = Proto.Property.Type.DATETIME;
            proto.datetime_value = property.Value.Ticks;
        }
    }
}

File: Java/src/main/java/org/thingmodel/proto/ToProtobuf.java
package org.thingmodel.proto;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

import org.thingmodel.Property.Location;
import org.thingmodel.PropertyType;
import org.thingmodel.ThingType;
import org.thingmodel.proto.ProtoProperty.Property;
import org.thingmodel.proto.ProtoStringDeclaration.StringDeclaration;
import org.thingmodel.proto.ProtoThing.Thing;
import org.thingmodel.proto.ProtoTransaction.Transaction;

public class ToProtobuf {
	// String declarations dictionary, for the sender side
	private HashMap<String, Integer> _stringDeclarations = new HashMap<>();
	
	// Current transaction object
	// It will be filled step by step
	private Transaction.Builder _transaction;
	
	private static HashMap<Class<? extends org.thingmodel.Property>, org.thingmodel.proto.ProtoPropertyType.PropertyType.Type>
		_prototypesBindings = new HashMap<>();
		
	static {
		_prototypesBindings.put(org.thingmodel.Property.Location.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.LOCATION);
		
		_prototypesBindings.put(org.thingmodel.Property.String.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.STRING);
		
		_prototypesBindings.put(org.thingmodel.Property.Double.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.DOUBLE);
		
		_prototypesBindings.put(org.thingmodel.Property.Integer.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.INT);
		
		_prototypesBindings.put(org.thingmodel.Property.Boolean.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.BOOLEAN);
		
		_prototypesBindings.put(org.thingmodel.Property.Date.class,
				org.thingmodel.proto.ProtoPropertyType.PropertyType.Type.DATETIME);
	}
	
	// When a string is in this collection, it should be sent
	// as a StringDeclaration for the next transaction
	private HashSet<String> _stringToDeclare = new HashSet<>();
	
	private HashMap<Integer, Thing.Builder> _thingsState = new HashMap<>();
	private HashMap<PropertyStateKey, Property.Builder> _propertiesState = new HashMap<>();

	private class PropertyStateKey {
		public int thingId;
		public int propertyId;
		
		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + propertyId;
			result = prime * result + thingId;
			return result;
		}
		
		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			PropertyStateKey other = (PropertyStateKey) obj;
			if (propertyId != other.propertyId)
				return false;
			if (thingId != other.thingId)
				return false;
			return true;
		}
		
	}
	
	protected int stringToKey(String value) {
		if (value == null || value.isEmpty()) {
			return 0;
		}
		
		Integer key = _stringDeclarations.get(value);
		if (key != null) {
			return key;
		}
		
		key = _stringDeclarations.size()+1;
		
		_stringDeclarations.put(value, key);

		_transaction.addStringDeclarations(
			StringDeclaration.newBuilder()
            	.setKey(key)
                .setValue(value));
		
		return key;
	}
	
	public Transaction Convert(Collection<org.thingmodel.Thing> publish,
			Collection<org.thingmodel.Thing> delete,
			Collection<ThingType> declarations,
			String senderID) {
		
		_transaction = Transaction.newBuilder();
		
		_transaction.setStringSenderId(stringToKey(senderID));
		
		for (org.thingmodel.Thing thing : publish) {
			ConvertThing(thing);
		}
		
		ConvertDeleteList(delete);
		ConvertDeclarationList(declarations);
		
		return _transaction.build();
	}
	
	protected void ConvertThing(org.thingmodel.Thing thing) {
		boolean change = false;
		
		int thingId = stringToKey(thing.getId());
		
		Thing.Builder publication = Thing.newBuilder()
			.setStringId(thingId)
			.setStringTypeName(thing.getType() != null ? stringToKey(thing.getType().getName()): 0)
			.setConnectionsChange(false);
		
		Thing.Builder previousThing = _thingsState.get(thingId);
		
		if (previousThing == null || previousThing.getStringTypeName() != publication.getStringTypeName()) {
			change = true;
		}
		
		List<org.thingmodel.Thing> connectedThingsCache = null;
		
		if ((previousThing == null && thing.getConnectedThingsCount() > 0)
			|| (previousThing != null && previousThing.getConnectionsCount() !=
				thing.getConnectedThingsCount())) {
			publication.setConnectionsChange(true);
		} else {
			connectedThingsCache = thing.getConnectedThings();
			
			for(org.thingmodel.Thing connectedThing : connectedThingsCache) {
				int connectionKey = stringToKey(connectedThing.getId());
				
				if (previousThing == null ||
						!previousThing.getConnectionsList().contains(connectionKey)) {
					publication.setConnectionsChange(true);
				}
			}
		}
		
		// If we don't have changes on the connection list
		// It's useless to send it
		if (publication.getConnectionsChange()) {
			change = true;
			
			if (connectedThingsCache == null) {
				connectedThingsCache = thing.getConnectedThings();
			}
			
			for (org.thingmodel.Thing connectedThing : connectedThingsCache) {
				int connectionKey = stringToKey(connectedThing.getId());
				
				publication.addConnections(connectionKey);
			}
		}
		
		for (org.thingmodel.Property property : thing.getProperties()) {
			int propertyId = stringToKey(property.getKey());
			
			Property.Builder proto = Property.newBuilder()
				.setStringKey(propertyId);
		
			Class<? extends org.thingmodel.Property> c = property.getClass();
			if (c == org.thingmodel.Property.Location.class) {
				ConvertLocationProperty((Location) property, proto);
			}
			else if (c == org.thingmodel.Property.String.class)
			{
				ConvertStringProperty((org.thingmodel.Property.String) property, proto);
			}
			else if (c == org.thingmodel.Property.Double.class)
			{
				proto.setType(Property.Type.DOUBLE);
				proto.setDoubleValue(((org.thingmodel.Property.Double) property).getValue());
			}
			else if (c == org.thingmodel.Property.Integer.class)
			{
				proto.setType(Property.Type.INT);
				proto.setIntValue(((org.thingmodel.Property.Integer) property).getValue());
			}
			else if (c == org.thingmodel.Property.Boolean.class)
			{
				proto.setType(Property.Type.BOOLEAN);
				proto.setBooleanValue(((org.thingmodel.Property.Boolean) property).getValue());
			}
			else if (c == org.thingmodel.Property.Date.class)
			{
				proto.setType(Property.Type.DATETIME);
				proto.setDatetimeValue(((org.thingmodel.Property.Date) property).getValue().getTime());
			}
			else if (c == org.thingmodel.Property.Double.class)
			{
				proto.setType(Property.Type.DOUBLE);
				proto.setDoubleValue(((org.thingmodel.Property.Double) property).getValue());
			}
			
			PropertyStateKey key = new PropertyStateKey();
			key.thingId = thingId;
			key.propertyId = propertyId;
			
			if (previousThing != null) {
				Property.Builder previousProto = _propertiesState.get(key);
				
				// Now, close your eyes !
				if (previousProto != null &&
					previousProto.getType().ordinal() == proto.getType().ordinal() &&
					previousProto.getBooleanValue() == proto.getBooleanValue() &&
					previousProto.getDatetimeValue() == proto.getDatetimeValue() &&
					previousProto.getDoubleValue() == proto.getDoubleValue() &&
					previousProto.getIntValue() == proto.getIntValue()
						) {
					
					Property.Location previousLoc = previousProto.getLocationValue();
					Property.Location loc = proto.getLocationValue();
					
					if ((previousLoc == null && loc == null) ||
						(previousLoc != null && loc != null &&
						previousLoc.getX() == loc.getX() &&
						previousLoc.getY() == loc.getY() &&
						previousLoc.getZ() == loc.getZ() &&
						previousLoc.getZNull() == loc.getZNull() &&
						previousLoc.getStringSystem() == loc.getStringSystem())) {
					
						Property.String previousStr = previousProto.getStringValue();
						Property.String str = proto.getStringValue();
						
						if ((previousStr == null && str == null) ||
							(previousStr != null && str != null &&
								((previousStr.getStringValue() ==  str.getStringValue() &&
								previousStr.getValue() == str.getValue()) ||
								(property.ValueToString().equals(previousStr.getValue()))))) {
							continue;
						}
					}
				}
			}
			
			change = true;
			
			publication.addProperties(proto);
			_propertiesState.put(key, proto);
			
		}
		
		if (change) {
			_transaction.addThingsPublishList(publication);
			_thingsState.put(thingId, publication);
		}
	}
	
	protected void ConvertDeleteList(Collection<org.thingmodel.Thing> list) {
		for (org.thingmodel.Thing thing : list) {
			int key = stringToKey(thing.getId());
			
			_transaction.addThingsRemoveList(key);
			ManageThingSuppression(key);
		}
	}
	
	protected void ConvertDeclarationList(Collection<org.thingmodel.ThingType> declarations) {
		for (ThingType thingType : declarations) {
			org.thingmodel.proto.ProtoThingType.ThingType.Builder declaration =
					org.thingmodel.proto.ProtoThingType.ThingType.newBuilder()
						.setStringName(stringToKey(thingType.getName()))
						.setStringDescription(stringToKey(thingType.Description));
			
			for (PropertyType propertyType : thingType.getProperties()) {
				declaration.addProperties(org.thingmodel.proto.ProtoPropertyType.PropertyType.newBuilder()
						.setStringKey(stringToKey(propertyType.getKey()))
						.setStringName(stringToKey(propertyType.Name))
						.setStringDescription(stringToKey(propertyType.Description))
						.setRequired(propertyType.Required)
						.setType(_prototypesBindings.get(propertyType.getType())));
			}
			
			_transaction.addThingtypesDeclarationList(declaration);
		}
	}
	
	protected void ConvertLocationProperty(org.thingmodel.Property.Location property,
			Property.Builder proto) {
		
		org.thingmodel.Location value = property.getValue();
		Class<? extends org.thingmodel.Location> c = value.getClass();
	
		if (c == org.thingmodel.Location.LatLng.class) {
			proto.setType(Property.Type.LOCATION_LATLNG);
		}
		else if (c == org.thingmodel.Location.Point.class) {
			proto.setType(Property.Type.LOCATION_POINT);
		}
		else {
			proto.setType(Property.Type.LOCATION_EQUATORIAL);
		}
	
		Property.Location.Builder loc = Property.Location.newBuilder()
				.setX(value.X)
				.setY(value.Y)
				.setStringSystem(stringToKey(value.System))
				.setZNull(value.Z == null);
		
		if (value.Z != null) {
			loc.setZ(value.Z);
		}
		
		proto.setLocationValue(loc);
	}
	
	protected void ConvertStringProperty(org.thingmodel.Property.String property,
			Property.Builder proto) {
		
		String value = property.getValue();
		proto.setType(Property.Type.STRING);
		
		Property.String.Builder st = Property.String.newBuilder();
		
		if (_stringToDeclare.contains(value)) {
			st.setStringValue(stringToKey(value));
		} else {
			st.setValue(value);
			
			// Use a string declaration next time
			_stringToDeclare.add(value);
		}

        proto.setStringValue(st);
		
	}
	
	protected void ManageThingSuppression(int thingId) {
		_thingsState.remove(thingId);
		
		ArrayList<PropertyStateKey> toRemove = new ArrayList<>();
		
		for (PropertyStateKey key : _propertiesState.keySet()) {
			if (key.thingId == thingId) {
				toRemove.add(key);
			}
		}
		
		for (PropertyStateKey key : toRemove) {
			_propertiesState.remove(key);
		}
	}
	
	public void ApplyThingsSuppressions(Collection<org.thingmodel.Thing> things) {
		for (org.thingmodel.Thing thing : things) {
			Integer key = _stringDeclarations.get(thing.getId());
			if (key != null){
				ManageThingSuppression(key);
			}
		}
	}
}
